In [1]:
import rasterio

In [62]:
dataset = rasterio.open('/data/1313.tif')

In [63]:
dataset.name

'/data/1313.tif'

In [64]:
dataset.mode

'r'

In [65]:
dataset.closed

False

Number of Bands in the Raster file

In [66]:
dataset.count

4

In [67]:
dataset.width

20084

In [68]:
dataset.height

64231

In [69]:
{i: dtype for i, dtype in zip(dataset.indexes, dataset.dtypes)}

{1: 'uint8', 2: 'uint8', 3: 'uint8', 4: 'uint8'}

In [70]:
dataset.bounds

BoundingBox(left=293966.86063, bottom=2276988.57419, right=294473.17827000003, top=2278607.8377)

World from 293966.86063 to 294473.1 meters, left to right
2276988.57419 to 2278607.8377 bottom to top.

In [11]:
dataset.transform

Affine(0.025210000000000003, 0.0, 293966.86063,
       0.0, -0.025210000000000003, 2278607.8377)

In [12]:
dataset.crs

CRS.from_epsg(32643)

https://www.spatialreference.org/ref/epsg/wgs-84-utm-zone-43n/

WGS 84 / UTM zone 43N is a projected CRS last revised on 17 June 2019 and is suitable for use in Between 72°E and 78°E, northern hemisphere between equator and 84°N, onshore and offshore. China. India. Kazakhstan. Kyrgyzstan. Maldives. Pakistan. Russian Federation. Tajikistan. WGS 84 / UTM zone 43N uses the WGS 84 geographic 2D CRS as its base CRS and the UTM zone 43N (Transverse Mercator) as its projection. WGS 84 / UTM zone 43N is a CRS for Large and medium scale topographic mapping and engineering survey.

https://georepository.com/crs_32643/WGS-84-UTM-zone-43N.html
https://epsg.io/32643

![](https://maps.tilehosting.com/styles/streets/static/auto/265x215@2x.png?key=qrAJy6x3Ck8n4XFFH4PS&latlng=1&fill=rgba(255,0,0,0.15)&stroke=red&width=2&path=0.0,72.0|84.0,72.0|84.0,78.0|0.0,78.0|0.0,72.0)

# Reading Raster Data

In [13]:
dataset.indexes

(1, 2, 3, 4)

Docker Container exited with code 137

https://www.petefreitag.com/item/848.cfm

Increase Memory allocated to Docker in Mac

In [14]:
band1 = dataset.read(1)

In [15]:
band1[0, 0]

0

In [19]:
band1[dataset.height//2, dataset.width//2]

93

In [20]:
dataset.bounds

BoundingBox(left=293966.86063, bottom=2276988.57419, right=294473.17827000003, top=2278607.8377)

In [26]:
import rasterio.warp
import rasterio.features
mask = dataset.dataset_mask()
for geom, val in rasterio.features.shapes(
        mask, transform=dataset.transform):

    # Transform shapes from the dataset's own coordinate
    # reference system to CRS84 (EPSG:4326).
    geom = rasterio.warp.transform_geom(
        dataset.crs, 'EPSG:4326', geom, precision=6)

    # Print GeoJSON shapes to stdout.
    print(geom)

/usr/local/lib/python3.6/site-packages/ipykernel_launcher.py:3: NodataShadowWarning: The dataset's nodata attribute is shadowing the alpha band. All masks will be determined by the nodata attribute
  This is separate from the ipykernel package so we can avoid doing imports until


{'type': 'Polygon', 'coordinates': [[[73.023133, 20.59528], [73.023321, 20.580657], [73.028177, 20.580712], [73.027989, 20.595335], [73.023133, 20.59528]]]}


In [27]:
geom.xy(geom.height//2, dataset.width//2)

AttributeError: 'dict' object has no attribute 'xy'

In [28]:
mask

array([[255, 255, 255, ..., 255, 255, 255],
       [255, 255, 255, ..., 255, 255, 255],
       [255, 255, 255, ..., 255, 255, 255],
       ...,
       [255, 255, 255, ..., 255, 255, 255],
       [255, 255, 255, ..., 255, 255, 255],
       [255, 255, 255, ..., 255, 255, 255]], dtype=uint8)

In [29]:
dataset.bounds

BoundingBox(left=293966.86063, bottom=2276988.57419, right=294473.17827000003, top=2278607.8377)

In [34]:
import time
start = time.time()
dataset.read(2)[dataset.height//2, dataset.width//2]
end = time.time()
runtime = end - start

In [35]:
start

1560801201.1029522

In [36]:
runtime

24.999762296676636

In [37]:
dataset.crs.wkt

'PROJCS["WGS 84 / UTM zone 43N",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Transverse_Mercator"],PARAMETER["latitude_of_origin",0],PARAMETER["central_meridian",75],PARAMETER["scale_factor",0.9996],PARAMETER["false_easting",500000],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["EPSG","32643"]]'

https://rasterio.readthedocs.io/en/stable/topics/georeferencing.html

In [38]:
dataset.transform

Affine(0.025210000000000003, 0.0, 293966.86063,
       0.0, -0.025210000000000003, 2278607.8377)

In [32]:
import timeit

In [39]:
from affine import Affine

Affine.identity()

Affine(1.0, 0.0, 0.0,
       0.0, 1.0, 0.0)

In [40]:
Affine.translation(1.0, 5.0)

Affine(1.0, 0.0, 1.0,
       0.0, 1.0, 5.0)

In [41]:
Affine.scale(2.0)

Affine(2.0, 0.0, 0.0,
       0.0, 2.0, 0.0)

In [42]:
Affine.shear(45.0, 45.0)

Affine(1.0, 0.9999999999999999, 0.0,
       0.9999999999999999, 1.0, 0.0)

In [43]:
Affine.rotation(45.0)

Affine(0.7071067811865476, -0.7071067811865475, 0.0,
       0.7071067811865475, 0.7071067811865476, 0.0)

In [45]:
dataset.height

64231

In [46]:
dataset.width

20084

In [47]:
left = 73.02313296121194
right = 73.02817687312424

In [48]:
width_per_pixel = (right - left)/dataset.width

In [49]:
width_per_pixel

2.511408042369839e-07

In [50]:
point = 73.02434934834984

In [51]:
distance = (point - left)

In [52]:
pixel_point = distance/width_per_pixel

In [53]:
pixel_point

4843.44685284117

In [55]:
round(pixel_point)

4843

In [57]:
band1[4843, 4842]

0

In [58]:
band1

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=uint8)

In [59]:
band2 = dataset.read(2)

In [60]:
band2[4843, 4842]

0

In [61]:
band2

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=uint8)

In [71]:
band1 = band2 = 0

In [72]:
band1 = dataset.read(2)

In [73]:
band1[4843, 4842]

0